### PR TITLE
Cleanup union and unionall

### PR DESCRIPTION
**Pull request by jingjingwang**
Based on the discussion in https://github.com/uwescience/raco/issues/484. `Union` is designed for sets and `UnionAll` is designed for bags. We use "+" for `UnionAll` and the word "Union" for `Union`. Clean up all the relevant places, also add an optimization rule to flatten a chain of (binary) `UnionAll`s.

### CODE REVIEW

**Comment by jingjingwang** [comment]
@bmyerz I'm not familiar with the grappa/clang side, I tried to find all the places that need to be modified, hopefully they make sense. Let me know if I'm missing anything!

**Comment by bmyerz** [comment]
What is the reasoning for removing `UNIONALL` syntax? This breaks some SQL compatibility. If there is no good reason, can you put in both `+` and `UNIONALL`?

**Comment by jingjingwang** [comment]
Since it's been a while, I rebased my commits on top of master. I added changes to kept both UNIONALL and "+".

File: raco/algebra.py [hunk 1]
```diff
@@ -421,23 +421,37 @@ def shortStr(self):
...
 
     """Bag union."""
 
-    def __init__(self, left=None, right=None):
-        BinaryOperator.__init__(self, left, right)
+    def __init__(self, children=None):
+        NaryOperator.__init__(self, children)
+
+    def partitioning(self):
+        return RepresentationProperties()
```
Comment (bmyerz) [inline raco/algebra.py]: Is this overly conservative? If all relations are identically partitioned can we keep the partitioning?

File: raco/backends/cpp/cpp.py [hunk 2]
```diff
@@ -548,7 +552,7 @@ def clangify(emit_print):
         rules.OneToOne(algebra.Project, CProject),
         rules.OneToOne(algebra.UnionAll, CUnionAll),
         # TODO: obviously breaks semantics
-        rules.OneToOne(algebra.Union, CUnionAll),
+        rules.OneToOne(algebra.Union, CUnion),
```
Comment (bmyerz) [inline raco/backends/cpp/cpp.py]: I'd prefer to leave Union as not implemented error

**Comment by jingjingwang** [comment]
@bmyerz Done! I made c tests use unionall, also fixed the partitioning of UnionAll. For Datalog tests, I made the IDBs use UnionAll directly, since it's compiled to the physical UnionAll operator anyway.

File: raco/algebra.py [hunk 3]
```diff
@@ -421,23 +421,46 @@ def shortStr(self):
...
+    def partitioning(self):
+        """keep the partitioning if all children are identically
+        partitioned"""
+        for child in self.args:
+            if child.partitioning.hash_partitioned != \
+                    self.args[0].partitioning.hash_partitioned:
+                return RepresentationProperties()
+        return self.args[0].partitioning
+
+    def partitioning(self):
```
Comment (bmyerz) [inline raco/algebra.py]: ill delete this duplicate

**Pull request merged**